### PR TITLE
[Example]rgcn-ogbn-mag

### PR DESCRIPTION
**Pull request by zyj-111**
## Description
There are two differents code examples of RGCN_ogbn-mag：
https://github.com/dmlc/dgl/tree/master/examples/pytorch/ogb/ogbn-mag
https://github.com/dmlc/dgl/tree/master/examples/pytorch/rgcn-hetero-ogbn-mag
This pr is going to merge two examples and replace some modules with existing modules in dgl.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- [x] [RelGraphEmbed](https://github.com/dmlc/dgl/blob/master/examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py#L21)-->[HeteroEmbedding](https://docs.dgl.ai/generated/dgl.nn.pytorch.HeteroEmbedding.html#dgl.nn.pytorch.HeteroEmbedding) in dgl
- [x] [add_reverse_hetero](https://github.com/dmlc/dgl/blob/master/examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py#L311)-->[AddReverse](https://docs.dgl.ai/generated/dgl.transforms.AddReverse.html) in dgl
- [x] merge two examples

### CODE REVIEW

**Comment by mufeili** [comment]
Could you also add a link to this example by updating the "Modeling Relational Data with Graph Convolutional Networks" item in the [README file](https://github.com/dmlc/dgl/blob/master/examples/README.md)?

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 1]
```diff
@@ -308,45 +286,9 @@ def prepare_data(args):
...
-
-        # copy_ndata:
-        for ntype in g.ntypes:
-            for k, v in g.nodes[ntype].data.items():
-                new_g.nodes[ntype].data[k] = v.detach().clone()
-
-        return new_g
-
-    g = add_reverse_hetero(g)
+    transform = AddReverse()
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:289]: You need to [compose](https://docs.dgl.ai/generated/dgl.transforms.Compose.html) AddReverse and [ToSimple](https://docs.dgl.ai/generated/dgl.transforms.ToSimple.html) as the original implementation invokes `to_simple`.

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 2]
```diff
@@ -14,44 +16,20 @@ def extract_embed(node_embed, input_nodes):
...
-    def forward(self, block=None):
-        return self.embeds
+def RelGraphEmbed(graph,embed_size,exclude):
+    node_num={}
+    for ntype in graph.ntypes:
+        if ntype in exclude:
+            continue
+        node_num[ntype] = graph.number_of_nodes(ntype)
+    embeds = HeteroEmbedding(node_num,embed_size)
+    for node_type in embeds.embeds.keys():
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:30]: You can directly add a `reset_parameters` function to the built-in `HeteroEmbedding` module.

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 3]
```diff
@@ -14,44 +16,20 @@ def extract_embed(node_embed, input_nodes):
     emb = {}
     for ntype, nid in input_nodes.items():
         nid = input_nodes[ntype]
-        if ntype in node_embed:
-            emb[ntype] = node_embed[ntype][nid]
+        if ntype in node_embed.embeds:
+            emb[ntype] = node_embed.embeds.pop(ntype)(nid)
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:20]: Do we need to "pop" it? Perhaps it's enough to do `node_embed.embeds[ntype]`?

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 4]
```diff
@@ -14,44 +16,20 @@ def extract_embed(node_embed, input_nodes):
...
-
-        self.reset_parameters()
-
-    def reset_parameters(self):
-        for emb in self.embeds.values():
-            nn.init.xavier_uniform_(emb)
-
-    def forward(self, block=None):
-        return self.embeds
+def RelGraphEmbed(graph,embed_size,exclude):
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:23]: For function names, typically we will use "rel_graph_embed" instead.

Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:23]: Add empty space after ",".

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 5]
```diff
@@ -14,44 +16,20 @@ def extract_embed(node_embed, input_nodes):
...
+    node_num={}
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:24]: "num={}" -> "num = {}"

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 6]
```diff
@@ -14,44 +16,20 @@ def extract_embed(node_embed, input_nodes):
...
+    for ntype in graph.ntypes:
+        if ntype in exclude:
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:26]: Perhaps you can directly do if `ntype == 'paper'` and remove the `exclude` argument.

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 7]
```diff
@@ -14,44 +16,20 @@ def extract_embed(node_embed, input_nodes):
...
+            continue
+        node_num[ntype] = graph.number_of_nodes(ntype)
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:28]: "number_of_nodes" -> "num_nodes" for brevity.

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 8]
```diff
@@ -14,44 +16,20 @@ def extract_embed(node_embed, input_nodes):
...
+    embeds = HeteroEmbedding(node_num,embed_size)
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:29]: Add empty space after ","

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py (hunk 3 above)
Comment (zyj-111) [reply examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:20]: The proper code may be "node_embed.embeds[ntype]（nid）“。The test results are as follows:
image

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py (hunk 8 above)
Comment (zyj-111) [reply examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:29]: Thank you for your suggestions above.

**Comment by mufeili** [comment]
Could you add a unit test for `reset_parameters` in [this function](https://github.com/dmlc/dgl/blob/5ba5106acab6a642e9b790e5331ee519112a5623/tests/pytorch/test_nn.py#L1257)?

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py [hunk 9]
```diff
@@ -17,18 +17,16 @@ def extract_embed(node_embed, input_nodes):
...
     for ntype in graph.ntypes:
-        if ntype in exclude:
+        if ntype == 'paper':
             continue
-        node_num[ntype] = graph.number_of_nodes(ntype)
-    embeds = HeteroEmbedding(node_num,embed_size)
-    for node_type in embeds.embeds.keys():
-        nn.init.xavier_uniform_(embeds.embeds[node_type].weight)  
+        node_num[ntype] = graph.num_nodes(ntype)
+    embeds = HeteroEmbedding(node_num, embed_size)
```
Comment (mufeili) [inline examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:29]: I think you need to add `embeds.reset_parameters()`.

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py (hunk 3 above)
Comment (mufeili) [reply examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py]: Yes. That's what I meant.

File: examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py (hunk 9 above)
Comment (zyj-111) [reply examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:29]: `embeds.reset_parameters()` is in the fuction `main` ,Line452.

Comment (mufeili) [reply examples/pytorch/ogb/ogbn-mag/hetero_rgcn.py:29]: Got it. Thanks.

**Comment by mufeili** [comment]
Replace `embed_layer()` with `embed_layer`

**Comment by mufeili** [comment]
LGTM in general. Let me make some final edit.

**Review by mufeili — APPROVED** [review]

**Pull request merged**